### ISSUE TITLE
A question about some code details

### ISSUE THREAD

**Issue by opposj** [maorash/NeRN#25]
I am confused by the following line of code:
https://github.com/maorash/NeRN/blob/879593d73be6166ca7fee09daab5dad2625a112d/NeRN/models/model.py#L190

I don't know why not assign value directly like this:
`curr_layer_weights.data = sampled_predicted_weights `

By the way, the utilization of distillation loss and attention loss seems to be invalid due to this function:
https://github.com/maorash/NeRN/blob/879593d73be6166ca7fee09daab5dad2625a112d/NeRN/trainer.py#L54

where gradients for the reconstructed model are disabled totally and I cannot find some code snippets that activate the `requires_grad`  attribute again. 

I think the most probable case is that the assignment of value can simultaneously provide `requires_grad=True` to `curr_layer_weights`, while ensuring some correct gradient flow (I'm not sure, it may be due to my *Pytorch* version, which is *1.13.1*, that I cannot observe such case).

**Comment by maorash** [comment]
Thanks for the comment. 

Regarding the weight value assignment - you're right. It's a part of some old code that allowed for gradient flow but is not relevant anymore. I've updated to code accordingly.

Regarding the gradients of the reconstructed model - we do not need them. The only relevant gradients are the ones with respect to NeRN's weights (the predictor). The reconstructed model is only used to compute the losses, and we initialize it using the weights predicted by NeRN at every iteration. We only update NeRN's weights, not the reconstructed model's.

**Comment by opposj** [comment]
Thank you for your response!

One of my puzzles, with respect to the weight assignment, has been resolved. However, I still have doubts about the validity of distillation loss.

I tried to set a breakpoint after the weight assignment:
![4062940da31660af30a87e9d5324b87](https://user-images.githubusercontent.com/49869100/216022798-03864153-5c78-413a-bc3d-e2f81a7dbc2f.png)

And this is what I obtained:
![2b0931ffcf1ccb81320d14120c969e2](https://user-images.githubusercontent.com/49869100/216023402-6a0df888-d708-4c85-80a6-edc1bb0b92af.png)

There is no `grad_fn` for `curr_layer_weights`. As a result, any convolution of the reconstructed model, which is related to `curr_layer_weights`, will not deliver gradients to the reconstructed weight, and thus the predictor. The distillation losses are closely related to the results of convolutions of the reconstructed model. I think such losses would not affect the optimization of the predictor.

In order to activate the gradient flow from the distillation losses to the predictor, I use the following code, where the reconstructed weights are directly used in convolution, rather than being assigned to parameters of `Conv2d` layers:
~~~
def update_weights(self, reconstructed_weights: List[torch. Tensor]):
    learnable_weights = self.get_learnable_weights()

    def cal_sampled_weights():
        sampled_weights = []
        for curr_layer_weights, curr_predicted_weights in zip(learnable_weights, reconstructed_weights):
            # Assuming a square filter
            curr_learnable_kernel_size = curr_layer_weights.shape[-1]
            curr_predicted_kernel_size = curr_predicted_weights.shape[-1]
            if self.sampling_mode == "center":
                min_coord = int(curr_predicted_kernel_size / 2 - curr_learnable_kernel_size / 2)
                max_coord = int(curr_predicted_kernel_size / 2 + curr_learnable_kernel_size / 2)
                sampled_predicted_weights = curr_predicted_weights[:, :, min_coord:max_coord,
                                            min_coord:max_coord]
            elif self.sampling_mode == "average":
                sampled_predicted_weights = F.avg_pool2d(curr_predicted_weights,
                                                         curr_predicted_kernel_size - curr_learnable_kernel_size + 1,
                                                         1)
            elif self.sampling_mode == "max":
                sampled_predicted_weights = F.max_pool2d(curr_predicted_weights,
                                                         curr_predicted_kernel_size - curr_learnable_kernel_size + 1,
                                                         1)
            else:
                raise ValueError(f"Unsupported sampling mode {self.sampling_mode}")
            sampled_weights.append(sampled_predicted_weights)
        return sampled_weights

    sampled_weights_list = cal_sampled_weights()

    for _i, cur_learnable_weight in enumerate(learnable_weights):
        for _layer in self.reconstructed_model.modules():
            if isinstance(_layer, torch.nn.Conv2d) and \
                    _layer.weight.shape == cur_learnable_weight.shape and \
                    torch.allclose(_layer.weight, cur_learnable_weight):

                def prevent_closure_modified_forward(_j):
                    def modified_forward(self, input: Tensor) -> Tensor:
                        return self._conv_forward(input, sampled_weights_list[_j], self.bias)

                    return modified_forward

                _layer.forward = prevent_closure_modified_forward(_i).__get__(_layer, torch.nn.Conv2d)

                break
~~~

After the adaptation, the training of `NeRN` suffered from severe gradient explosion and the loss would turn to `NaN` in no time. I think there are still some problems w.r.t. the distillation losses.

**Comment by maorash** [comment]
Hi, sorry for the late response, your comment did not re-open the issue. 

It looks like the implementation we shared does indeed have a problem with the distillation losses. We have done a great deal of refactoring to make the code usable and readable before sharing it. One of these refactors was trying to handle the gradient flow in a more graceful fashion, but we didn’t validate it well enough.

We've fixed this issue and made sure the gradients are handled properly. We are currently running some additional experiments to make sure there are no bugs left behind and will update the code as soon as we're done - making sure the experiments are easily reproducible. We'll also do our best to provide the pretrained weights.

Thanks again for being observant and bringing this to our attention 😄

**Comment by opposj** [comment]
Cool! Looking forward to the upcoming update of the code :)

**Comment by hchoi71** [comment]
Is this issue related to distillation loss resolved in the current code?

**Comment by hchoi71** [comment]
> Cool! Looking forward to the upcoming update of the code :)

Is this issue related to distillation loss resolved in the current code?
